Add lambda defaults/`params` tests from the test plan review

jjonescz opened the pull request:
Adding tests from [the test plan](https://github.com/dotnet/roslyn/issues/62485) as discussed during the test plan review meeting, except for NoPIA which I didn't figure out how to test.

333fred commented:
> except for NoPIA which I didn't figure out how to test

Alright, NoPIA is a bit of a pain to test. PIAs are about embedding interop types. An assembly that sets embed interop types says that if a downstream assembly references a type from it, it needs to actually copy the type into that assembly. You can test it by creating a test with two compilations in it: The first assembly defines some types that the second will use. When referencing the first assembly from the second assembly, you call `EmitToImageReference().WithEmbeddedInteropTypes(true)`. On the second assembly, you then want to call `CompileAndVerify`, and pass a `symbolValidator` to validate that the emitted module symbol did actually embed the type you were expecting in the second compilation. In terms of lambdas, there are two scenarios that need to be tested:

1. The positive scenario, when you do something like `var l = (EmbeddedType t = null) => {};`, which should cause `EmbeddedTpe` to be embedded.
2. The negative scenario, when you do something like `var l = (string s = EmbeddedType.ConstantStringField) => {};`. Because the constant value should be copied, `EmbeddedType` should not actually embedded here.

I'll be out next week, so if you run into issues with this you can talk to @AlekseyTs.

333fred reviewed (commented):
Done review pass (commit 1)

jjonescz commented on src/Compilers/CSharp/Test/Semantic/Semantics/DelegateTypeTests.cs:
```diff
@@ -13337,6 +13349,9 @@ public static void Main()
         [InlineData("char", "'a'", "a")]
         [InlineData("string", @"""a string""", "a string")]
         [InlineData("C", "null", "")]
+        [InlineData("C", "default(C)", "")]
```
Thanks, added: 64f858d85ba2bd5fccf38894f00b96882ae81d30

jjonescz commented:
Thanks, @333fred, I added NoPIA tests: 384c934db6e627a048fb72674fb74824b0477dbf

However, it seems impossible to test `var l = (string s = EmbeddedType.ConstantStringField) => {};`, because `EmbeddedType` must be an interface and if it has a constant member, this error is raised:

```
error CS8711: Type 'EmbeddedType' cannot be embedded because it has a non-abstract member. Consider setting the 'Embed Interop Types' property to false.
```

So I added similar negative scenario instead: `var l = (object o = default(EmbeddedType)) => {};`
